Assets/Scripts/Gameplay/GameplayObjects/Breakable.cs

```diff
@@ -0,0 +1,206 @@
+                }
+            }
+
+            if (IsClient)
+            {
+                IsBroken.OnValueChanged += OnBreakableStateChanged;
+
+                if (IsBroken.Value == true)
+                {
+                    PerformBreakVisualization(true);
+                }
+            }
+        }
+
+        public override void OnNetworkDespawn()
+        {
+            if (IsClient)
+            {
+                IsBroken.OnValueChanged -= OnBreakableStateChanged;
+            }
+        }
+
+        public void ReceiveHP(ServerCharacter inflicter, int HP)
+        {
+            if (HP < 0)
+            {
+                if (inflicter && !inflicter.IsNpc)
+                {
+                    bool isNotDamagedByPlayers = (GetSpecialDamageFlags() & IDamageable.SpecialDamageFlags.NotDamagedByPlayers) == IDamageable.SpecialDamageFlags.NotDamagedByPlayers;
+                    if (isNotDamagedByPlayers)
+                    {
+                        // a player tried to damage us, but we are immune to player damage!
+                        return;
+                    }
+                }
+
+                if (m_NetworkHealthState)
+                {
+                    m_NetworkHealthState.HitPoints.Value =
+                        Mathf.Clamp(m_NetworkHealthState.HitPoints.Value + HP, 0, m_MaxHealth.Value);
+                    if (m_NetworkHealthState.HitPoints.Value <= 0)
+                    {
+                        Break();
+                    }
+                }
+                else
+                {
+                    //any damage at all is enough to slay me.
+                    Break();
+                }
+            }
+        }
+
+        private void Break()
+        {
+            IsBroken.Value = true;
+            if (m_Collider)
+                m_Collider.enabled = false;
+        }
+
+        public void Unbreak()
+        {
+            IsBroken.Value = false;
+            if (m_Collider)
+                m_Collider.enabled = true;
+            if (m_MaxHealth && m_NetworkHealthState)
+                m_NetworkHealthState.HitPoints.Value = m_MaxHealth.Value;
+        }
+
+        public IDamageable.SpecialDamageFlags GetSpecialDamageFlags()
+        {
+            return m_SpecialDamageFlags;
+        }
+
+        public bool IsDamageable()
+        {
+            // you can damage this breakable until it's broken!
+            return !IsBroken.Value;
+        }
+
+        private void OnBreakableStateChanged(bool wasBroken, bool isBroken)
+        {
+            if (!wasBroken && isBroken)
+            {
+                PerformBreakVisualization(false);
+            }
+            else if (wasBroken && !isBroken)
+            {
+                PerformUnbreakVisualization();
+            }
+        }
+
+        private void PerformBreakVisualization(bool onStart)
+        {
+            foreach (var gameObject in m_UnbrokenGameObjects)
```

Comment:
nit: We could have a few asserts to make sure those methods are running client or server side and make sure this is invariants?

Reply:
This code path goes from a place where we check for `IsClient`, however I'm not against adding `Assert.IsTrue(IsClient)` in some places, but again - in another PR.